### PR TITLE
deepchem pytorch lightning tutorial

### PR DESCRIPTION
**Pull request by Chahalprincy**
## Description
This pull request implements PyTorch lightning implementation for deepchem model.

List any dependencies that are required for this change. -->
PyTorch lightning https://www.pytorchlightning.ai/

## hydra + lightning + deepchem

For a future feature we can also explore the integration of hydra (https://github.com/facebookresearch/hydra) with the lightning + deepchem port. I have not included that functionality here because the power of hydra really shines through when running experiments from the command line and not inside notebooks.

Notes on hydra:
- Hydra has available integration with pytorch lightning (https://github.com/ashleve/lightning-hydra-template) which we can build upon.
- Hydra provides functionality for building ML modules and running ML experiments.
- It has a seamless integration with SLURM through [submitit](https://github.com/facebookincubator/submitit) which makes hyperparameter searches easier.

## DGL dependency

This tutorial also answers a question related to DGL discussed in https://github.com/deepchem/deepchem/issues/2594. In this tutorial we have used a GCN (Graph Convolutional Network) which underneath uses DGL. This example works well so potentially there are no serious conflicts with using pytorch lightning with DGL at least.

## Deepchem architecture

There are some conflicts around the lightning modules and deepchem's pytorch models, specification of learning rate, batch size etc. Currently I have shown a proof of concept in this tutorial which writes some boilerplate wrapper code to make things work. Going forward we should think about what would be a good way to integrate these two APIs.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by rbharath** [comment]
We did a quick review on the call with Abhishek, but to recap here, the main changes needed are more expository text and explanations of the material here. Looking forward to reviewing a revised version!

**Comment by Chahalprincy** [comment]
Thanks!! @rbharath, I am working on the changes and proposal. I will push it out soon.

**Comment by Chahalprincy** [comment]
Hi @rbharath , I have updated the PR, please let me know if any further changes are required.

File: examples/tutorials/PytorchLightning_Integration.ipynb [hunk 1]
```diff
@@ -0,0 +1,464 @@
+{
```
Comment (rbharath) [inline examples/tutorials/PytorchLightning_Integration.ipynb:1]: Should add tutorial on to examples/tutorials/README.md so users know about it.

File: examples/tutorials/PytorchLightning_Integration.ipynb [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "6890f523",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import deepchem as dc\n",
```
Comment (rbharath) [inline examples/tutorials/PytorchLightning_Integration.ipynb:47]: We should add the installation commands such as
```
pip install --pre deepchem
pip install pytorch_lightning
```

File: examples/tutorials/PytorchLightning_Integration.ipynb [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+    "\n",
+    "Below we show an example of a Graph Convolution Network (GCN). Note that this is a simple example which uses a GCNModel to predict the label from an input sequence. We do not showcase the complete functionality of deepchem in this example as we want to restructure the deepchem code and adapt it so that it can be easily plugged into pytorch-lightning. This example was inspired from the `GCNModel` documentation present [here](https://github.com/deepchem/deepchem/blob/a68f8c072b80a1bce5671250aef60f9cc8519bec/deepchem/models/torch_models/gcn.py#L200)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2859f97b",
+   "metadata": {},
+   "source": [
+    "**Prepare the dataset**"
```
Comment (rbharath) [inline examples/tutorials/PytorchLightning_Integration.ipynb:75]: For this and the following sections, we should have a sentence description of what we are doing for first time readers or new DeepChem users.

File: examples/tutorials/PytorchLightning_Integration.ipynb [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+    "        featurizer = dc.feat.MolGraphConvFeaturizer()\n",
+    "        X = featurizer.featurize(smiles)\n",
+    "        self._samples = dc.data.NumpyDataset(X=X, y=labels)\n",
+    "        \n",
+    "    def __len__(self):\n",
+    "        return len(self._samples)\n",
+    "        \n",
+    "    def __getitem__(self, index):\n",
+    "        return (\n",
+    "            self._samples._X[index],\n",
```
Comment (rbharath) [inline examples/tutorials/PytorchLightning_Integration.ipynb:200]: We should use
```
self._samples.X[index]
...
```

File: examples/tutorials/PytorchLightning_Integration.ipynb [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "44df652a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class GCNDatasetModule(pl.LightningDataModule):\n",
```
Comment (rbharath) [inline examples/tutorials/PytorchLightning_Integration.ipynb:233]: Perhaps rename to `SmilesDatasetModule` since we don't use GCN in this class

File: examples/tutorials/PytorchLightning_Integration.ipynb [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "bfff80bd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create module objects\n",
+    "gcn_datasetmodule = GCNDatasetModule(\n",
```
Comment (rbharath) [inline examples/tutorials/PytorchLightning_Integration.ipynb:333]: Perhaps useful to have a conversion function or example showing how to convert a DeepChem Dataset object into a DatasetModule object.

**Review by rbharath — COMMENTED** [review]
Tutorial is looking good. Once comments are addressed should be able to merge in

File: examples/tutorials/PytorchLightning_Integration.ipynb (hunk 6 above)
Comment (Chahalprincy) [reply examples/tutorials/PytorchLightning_Integration.ipynb]: Hi Bharath, it is tough to create a generic function here to convert any deepchem dataset object into a datasetmodule as I think it will need a good amount of wrapper implementation to handle the complexity. If it is okay I would like to keep the current design for the tutorial and we can implement a conversion function in the next PR.

**Comment by Chahalprincy** [comment]
Hey Bharath, I have updated the PR following up on the comments.

**Review by rbharath — APPROVED** [review]
LGTM
